_posts/2020-04-17-quarkus-mutual-tls.adoc

```diff
@@ -0,0 +1,405 @@
+}
+----
+
+Inject the GreetingService rest client on `org.acme.client.mtls.GreetingResource`.
+
+.org.acme.client.mtls.GreetingResource
+[source,java]
+----
+@Path("/hello")
+public class GreetingResource {
+
+    @Inject // <1>
+    @RestClient // <2>
+    GreetingService greetingService;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return greetingService.hello(); // <3>
+    }
+}
+----
+<1> CDI `@Inject` annotation.
+<2> MicroProfile `@RestClient` annotation.
+<3> Replace the return statement with the `greetingService.hello();`.
+
+NOTE: See the guide https://quarkus.io/guides/rest-client[rest-client] to explore in details.
+
+=== Update the unit test
+
+Add `quarkus-junit5-mockito` dependency to your project.
+
+.pom.xml
+[source,xml]
+----
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-junit5-mockito</artifactId>
+    </dependency>
+----
+
+.org.acme.client.mtls.GreetingResourceTest
+[source,java]
+----
+@QuarkusTest
+public class GreetingResourceTest {
+
+    @InjectMock // <1>
+    @RestClient // <2>
+    GreetingService greetingService;
+
+    @Test
+    public void testHelloEndpoint() {
+        Mockito.when(greetingService.hello()).thenReturn("hello from server"); // <3>
+
+        given()
+          .when().get("/hello")
+          .then()
+             .statusCode(200)
+             .body(is("hello from server"));
+    }
+
+}
+----
+<1> Inject the CDI bean.
+<2> RestClient type.
+<3> Mock the hello request.
+
+NOTE: See the guide https://quarkus.io/guides/getting-started-testing[Testing] to explore in details.
+
+=== Configure MicroProfile rest client for Mutual TLS.
+
+Add the following properties to enable SSL in your application `src/main/resources/application.properties`
+
+.application.properties
+[source]
+----
+org.acme.client.mtls.GreetingService/mp-rest/url=https://localhost:8443
+org.acme.client.mtls.GreetingService/mp-rest/trustStore=classpath:/META-INF/resources/client.truststore
+org.acme.client.mtls.GreetingService/mp-rest/trustStorePassword=password
+org.acme.client.mtls.GreetingService/mp-rest/keyStore=classpath:/META-INF/resources/client.keystore
+org.acme.client.mtls.GreetingService/mp-rest/keyStorePassword=password
+
+quarkus.ssl.native=true
+----
+
+=== Run it.
+
+[source]
+----
+mvn quarkus:dev
+----
+
+Now let's hit the client _/hello_ endpoint, and you should expect the following.
+
+[source]
+----
+curl http://localhost:8080/hello
+hello from server
+----
+
+== External Configuration
+
+You do not want to have a certificate inside your application and Quarkus allows you to use external configuration and override the runtime application properties.
+Using the external `config` directory under the current working directory, for instance `config/application.properties`
+
+Let's check how we can do in Kubernetes / OpenShift.
+
+[NOTE]
+====
+Working directory:
+
+* `redhat-openjdk-18/openjdk18-openshift` image is `/deployments`
```

Comment:
I don't understand this

Reply:
The JVM and native image have two different working directory.
 - jvm has the working directory under /deployments/
 - native under /home/quarkus
To be able to do properties override we need to mount the configMap on <working-directory>/config
We can show only one case if it is confusing.

Reply:
I still don't get it, sorry :/ I'd say this section needs some serious context explanation.
As a new reader, I have no idea why you have those two things nor what they represent really.

@geoand is there something we should improve on Quarkus?

Reply:
I see what @eye0fra is saying here. The point is that the two container images use different paths, therefore you need to use different values for the properties in order to "target" the appropriate keystore.

I am thinking that perhaps we can generate a ConfigMap automatically for the user containing this information (or maybe add it as env vars to the `Deployment` / `DeploymentConfig`) when TLS is enabled.

We should definitely explore this to make it as easy as possible for users. @eye0fra can you please open an issue in Quarkus explaining this case (you can even link to this discussion for more context) so we can explore it?

Thanks

Reply:
This is more how the s2i quarkus image is designed: 
https://github.com/quarkusio/quarkus-images/tree/master/modules/quarkus-native-s2i-scripts
And I think there is no way to override the QUARKUS_HOME variable with s2i at the moment.
I will open an issue on Quarkus to allow more flexibility.

Reply:
https://github.com/quarkusio/quarkus/issues/8820